File: src/Game/Entities/FireRune.cpp
#include "Game/Entities/FireRune.h"
#include "Game/Entities/Unit.h"
#include "Game/Entities/DamageInfo.h"
#include "Game/World.h"

FireRune::FireRune(World* world, int width, int height)
    : AreaTrigger(world, width, height) {
    SetName("Fire Rune");
    // cycle the three rune frames (image-based animation)
    GetAnimation()
        .addFrame(Sprite(SpriteId::FireRune0), 0.18f)
        .addFrame(Sprite(SpriteId::FireRune1), 0.16f)
        .addFrame(Sprite(SpriteId::FireRune2), 0.20f);
}

void FireRune::OnUpdate(float dt) {
    AreaTrigger::OnUpdate(dt);   // membership detection + enter/exit hooks

    tick_ += dt;
    if (tick_ < 1.0f) return;
    tick_ -= 1.0f;

    // burn everyone standing on the rune
    for (Unit* unit : GetWorld()->GetUnits()) {
        if (unit->IsAlive() && Contains(unit->GetX(), unit->GetY())) {
            DamageInfo dmg(damagePerSecond_, DamageSchool::Fire);
            unit->TakeDamage(dmg);
        }
    }
}

void FireRune::OnUnitEnter(Unit* unit) {
    GetWorld()->AddMessage(unit->GetName() + " steps into the fire rune!");
}

void FireRune::OnUnitExit(Unit* unit) {
    GetWorld()->AddMessage(unit->GetName() + " leaves the fire rune.");
}

File: scripts/entities/firerune.lua
return {
    name = "Fire Rune",
    width = 5,
    height = 3,
    damagePerSecond = 1,

    light = { r = 1.0, g = 0.55, b = 0.18, radius = 6.0, intensity = 1.7 },

    frames = {
        { sprite = Sprite.FireRune0, dur = 0.18 },
        { sprite = Sprite.FireRune1, dur = 0.16 },
        { sprite = Sprite.FireRune2, dur = 0.20 },
    },

    onUpdate = function(self, dt)
        self.tick = (self.tick or 0) + dt
        if self.tick < 1.0 then return end
        self.tick = self.tick - 1.0
        for _, u in ipairs(self.unitsInside()) do
            u:takeDamage(self.damagePerSecond, DamageSchool.Fire)
        end
    end,

    onEnter = function(self, unit)
        self.log(unit:name() .. " steps into the fire rune!")
    end,

    onExit = function(self, unit)
        self.log(unit:name() .. " leaves the fire rune.")
    end,
}
